RFCs/rfc-0008-wallet-provider-api.md

```diff
@@ -0,0 +1,387 @@
+
+### Connectivity
+
+The Provider is "connected" when it can service RPC requests to at least one chain.
+
+### API
+
+#### request
+
+The `request` method is a transport- and protocol-agnostic wrapper function for Remote Procedure Calls (RPCs).
+
+```typescript
+interface RequestArguments {
+  readonly method: string;
+  readonly params?: readonly unknown[] | object;
+}
+
+Provider.request(args: RequestArguments): Promise<unknown>;
+```
+
+### Supported RPC Methods
+
+A "supported RPC method" is any RPC method that may be called via the Provider.
+
+All supported RPC methods MUST be identified by unique strings. The unique identifiers must be consistent across all wallet providers.
+
+Providers MAY support whatever RPC methods required to fulfill their purpose, standardized or otherwise.
+
+If an RPC method defined in a finalized RFC (or MIP) is not supported, it SHOULD be rejected with a `4200` error per the Provider Errors section below, or an appropriate error per the RPC method's specification.
+
+### RPC Errors
+
+```typescript
+interface ProviderRpcError extends Error {
+  message: string;
+  code: number;
+  data?: unknown;
+}
+```
+
+#### Error Standards
+
+`ProviderRpcError` codes and messages should follow these conventions:ß◊
+- `4001`: User Rejected Request
+- `4100`: Unauthorized
+- `4200`: Unsupported Method
+- `4900`: Disconnected
+- `4901`: Chain Disconnected
+
+### Events
+
+The Provider **MUST** implement the following event handling methods:
+- `on` 
+- `removeListener` methods. 
+
+These methods MUST be implemented per the Node.js EventEmitter API.
+
+
+It also should handle the `mina_message` and `connect` events.
+
+#### mina_message
+
+This event is for arbitrary notifications. When emitted, the `mina_message` event must be with an object argument:
+
+```typescript
+interface ProviderMessage {
+  readonly type: string;
+  readonly data: unknown;
+}
+```
+
+#### connect
+
+The connect event must be emitted with an object:
+
+```typescript
+interface ProviderConnectInfo {
+  readonly chainId: string;
+}
+```
+
+#### disconnect
+
+The Provider is said to be "disconnected" when it cannot service RPC requests to any chain at all.
+
+If the Provider becomes disconnected from all chains, the Provider MUST emit the event named `disconnect` with value error: ProviderRpcError, per the interfaced defined in the RPC Errors section. The value of the error's code property MUST follow the status codes for CloseEvent.
+
+#### chainChanged
+
+If the chain the Provider is connected to changes, the Provider MUST emit the event named `chainChanged` with value `chainId: string`, specifying the integer ID of the new chain as a hexadecimal string, per the `mina_chainId` RPC method.
+
+#### accountsChanged
+
+If the accounts available to the Provider change, the Provider MUST emit the event named `accountsChanged` with value `accounts: string[]`, containing the account addresses per the `mina_accounts` RPC method.
```

Comment:
Switching to an account that is not available to the provider would mean the user has to reconnect, upon reconnecting that account becomes available and therefore the wallet must emit the `accountsChanged` event.

However if the user does not re-connect then no event should be triggered because there has not been a change to the accounts available.

Thoughts?